singlem/diamond_spkg_searcher.py

```diff
@@ -70,41 +72,57 @@ def _prefilter(self, diamond_database, read_files, is_reverse_reads, performance
-                  "--evalue 0.01 " \
-                  "%s " \
-                  "--threads %i " \
-                  "--query %s " \
-                  "--db %s " \
-                  "| tee >(sed 's/^/>/; s/\\t/\\n/; s/\\t.*//' > %s) " \
-                  "| awk '{print $1,$3}'" % (
-                      performance_parameters,
-                      self._num_threads,
-                      file,
-                      diamond_database,
-                      fasta_path)
-            logging.debug(cmd)
-
-            # Originially, we ran here via os.system rather than normal extern
-            # so reads can be piped in to singlem. However, this meant that
-            # errors and failed commands were ignored, sometimes causing
-            # successful return of singlem but an empty OTU table.
-            qseqid_sseqid = extern.run(cmd)
+
+            # DIAMOND command
+            # now with range culling, etc
+            cmd = [ 
+                "diamond", "blastx",
+                "--outfmt", "6", "qseqid", "full_qseq", "sseqid",
+                "--max-target-seqs", "0",
+                "--evalue", "0.01",
+                "--frameshift", "15",
+                "--range-culling",
+                "-k", "1",
+                "--threads", str(self._num_threads),
+                "--query", file,
+                "--db", diamond_database
+            ]
+
+            cmd.extend(performance_parameters.split())
+            logging.debug(' '.join(cmd))
+
             best_hits = {}
-            for line in qseqid_sseqid.splitlines():
-                try:
-                    (qseqid,sseqid) = line.split(' ')
-                except ValueError:
-                    raise Exception("Unexpected line format for DIAMOND output line '{}'".format(line))
-                if qseqid in best_hits and best_hits[qseqid] != sseqid:
-                    raise Exception("Multiple DIAMOND best hits detected for '{}'. This likely indicates that the input reads have non-unique names, possibly due to the same read appearing twice in a single input file".format(qseqid))
-                best_hits[qseqid] = sseqid
+            # using Popen to stream the output
+            with Popen(cmd, stdout=PIPE, stderr=PIPE, text=True) as proc:
+                with open(fasta_path, 'a') as fasta_file:
+                    for line in proc.stdout:
+                        try:
+                            qseqid, full_qseq, sseqid = line.strip().split('\t')
+                        except ValueError:
+                            raise Exception(f"Unexpected line format for DIAMOND output line '{line.strip()}'")
+                    
+                        # creating new read index to account for multiple hits
+                        qseqid = qseqid + '~' + sseqid.split('~')[0]
 
-            diamond_results.append(DiamondSearchResult(fasta_path, best_hits))
+                        # DIAMOND shouldn't return the same qseqid twice, but just in case
```

Comment:
change this comment as the original qseqid is reported more than once for nanopore